DOC: Fixed timedelta docstring and examples

ryankarlos opened the pull request:
- [x] closes #23248
- [x] tests added / passed
- [x] passes `git diff upstream/master --name-only -- "*.py" | grep "pandas/" | xargs flake8`
- [x] passes `python scripts/validate_docstrings.py pandas.to_timedelta`

ryankarlos committed:
Modified timedelta docstring to include M and Y units and examples

pep8speaks commented:
Hello @ryankarlos! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/tools/timedeltas.py`](https://github.com/ryankarlos/pandas/blob/072277c4bf63a7b53739ce47d3e58777e78eaef6/pandas/core/tools/timedeltas.py) !

ryankarlos commented:
I've added in a few examples to illustrate 'M' and 'Y' outputs along with generating new columns in a dataframe using timedelta. Also corrected some linting issues and added a better summary. Any feedback would be welcome.

ryankarlos commented:
Travis failing in feather --- ive seen other PRs having the same problem - not sure why it is ?

jreback commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,21 +18,27 @@
+
+    Timedeltas are differences in times, expressed in difference units,
+    for example, years, momths, days, hours, minutes, seconds.
+    They can be both positive and negative. This method can create Timedelta
```
this is not true at all

only Days and smaller are supoorted 

these are fixed time units

if you want variable units use an Offset which are designed for this

ryankarlos committed:
Corrected summary

mroeschke commented on pandas/core/tools/timedeltas.py:
```diff
@@ -64,6 +69,31 @@ def to_timedelta(arg, unit='ns', box=True, errors='raise'):
     TimedeltaIndex(['0 days', '1 days', '2 days', '3 days', '4 days'],
                    dtype='timedelta64[ns]', freq=None)
 
+    For `M` and `Y` units, `1M = 30D` and 1Y = 365D:
```
So there's an issue to deprecate `'M'` and `'Y'` for Timedelta, so I don't think it's a good idea to advertise these argument that are slated for deprecation. 

https://github.com/pandas-dev/pandas/issues/16344

ryankarlos replied:
@mroeschke  sorry wasn't aware about the issue - ill remove it. Im thinking giving example for all the other units would be overkill - maybe a couple more ? 
Maybe a reference to DateOffset in the description or examples would help ?

mroeschke replied:
If you're up for it, you could address #16344 first and then add a note here saying "Day frequencies and higher are only supported.

ryankarlos replied:
Just sent an initial PR for #16344 , still have work to do as its my first time fixing deprecation issues before

mroeschke commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,21 +18,26 @@
-    Convert argument to timedelta
+    Convert argument to timedelta.
+
+    Timedeltas are differences in times, expressed in difference units
```
Can you mention they are _absolute_ differences in time.

ryankarlos committed:
Added box example and removed M and Y unit references

mroeschke commented on pandas/core/tools/timedeltas.py:
```diff
@@ -64,6 +69,23 @@ def to_timedelta(arg, unit='ns', box=True, errors='raise'):
+    >>> pd.to_timedelta(np.arange(5), box=False)
+    array([0, 1, 2, 3, 4], dtype='timedelta64[ns]')
+
+    Add new column of dates from existing dates in a `DataFrame`
```
I'd remove this example as well; it demonstrates assignment more than `to_timedelta`

mroeschke replied:
The one below it demonstrating DataFrame column assignment

ryankarlos committed:
Removed last example

mroeschke commented:
cc @datapythonista if you could have a look.

ryankarlos commented:
I haven't added the notes section for deprecated units yet - i was working concurrently on #23264 so  wasn't sure if i should add that in before finishing that

datapythonista reviewed (commented):
Thanks for the work on this. Added some comments about the formatting.

ryankarlos commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,21 +18,26 @@
-          timedelta64[ns]
+    arg : String, timedelta, list, tuple, 1-d array, or Series
+          The argument which needs to be converted to timedelta.
+    unit : Integer or float, default ns
```
Oh Sorry yes - my bad - thanks for spotting this.

ryankarlos commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,21 +18,26 @@
+          Denotes the unit (D,h,m,s,ms,us,ns) of the arg.
+    box : Boolean, default True
+          If True returns a Timedelta/TimedeltaIndex of the results.
+          if False returns a np.timedelta64 or ndarray of values of dtype
```
Thanks, thats why i removed them below to get rid of the validation errors - will include them.

ryankarlos committed:
improved styling

ryankarlos commented:
@datapythonista Thanks for the review. Added changes. Also rendered the docstring below:

![to_timedelta_docstring](https://user-images.githubusercontent.com/16509490/47596584-35b61700-d97f-11e8-912e-e6b3e6a08218.png)

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,26 +18,36 @@
+    Convert argument to timedelta.
+
+    Timedeltas are absolute differences in times, expressed in difference
+    units e.g. days, hours, minutes, seconds. This method converts an argument
```
I think it'd look better to have the `e.g.` in brackets.

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,26 +18,36 @@
+    arg : str, timedelta, list, tuple, 1-d array, or Series
+          The argument which needs to be converted to timedelta.
+    unit : str, default 'ns'
+          Denotes the unit (D,h,m,s,ms,us,ns) of the arg.
```
I'd add spaces after the commas.

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,26 +18,36 @@
 
     Returns
     -------
     ret : timedelta64/arrays of timedelta64 if parsing succeeded
```
you can get rid of the `ret`, and leave here just the types `timedelta64 or numpy.array of timedelta64`. We should be able to automatically parse the line with the types, so no description should go there. A description on what is being returned, and clarification on when every type is being returned should go in the next line (indented).

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,26 +18,36 @@
+          Denotes the unit (D,h,m,s,ms,us,ns) of the arg.
+    box : bool, default True
+        - If True returns a Timedelta/TimedeltaIndex of the results.
         - if False returns a np.timedelta64 or ndarray of values of dtype
```
Capital `I`

datapythonista reviewed (commented):
looking good, added a few more comments.

ryankarlos committed:
styling/linting

ryankarlos commented:
@datapythonista Thanks, added an update

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -21,18 +21,18 @@ def to_timedelta(arg, unit='ns', box=True, errors='raise'):
     Parameters
     ----------
-    arg : str, timedelta, list, tuple, 1-d array, or Series
+    arg : str,  timedelta,  list,  tuple,  1-d array,  or Series
```
the missing spaces were down, in the `(D,h,m,s...)`, which should be `(D, h, m, s,...)`.

Just leave a space after the commas here.

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,25 +18,36 @@
+    Convert argument to timedelta.
+
+    Timedeltas are absolute differences in times, expressed in difference
+    units e.g. (days, hours, minutes, seconds). This method converts an argument
```
I'd move the `e.g.` inside the parenthesis  I think that makes more sense.

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -18,25 +18,36 @@
-        - If True returns a Timedelta/TimedeltaIndex of the results
-        - if False returns a np.timedelta64 or ndarray of values of dtype
-          timedelta64[ns]
+    arg : str, timedelta, list, tuple, 1-d array, or Series
```
I'd replace list, tuple, 1-d array by `list-like`. I'd also remove the comma vefore the or.

datapythonista reviewed (commented):
Just spotted couple of last things. If the examples are now working, looks good to me after these last fixes.

ryankarlos committed:
Final fixes and Fixed merge conflict/included recent upstream changes

ryankarlos commented:
@datapythonista done :)
![screen shot 2018-10-27 at 15 05 57](https://user-images.githubusercontent.com/16509490/47605088-df39ee80-d9f9-11e8-8fd1-acea5ad8760f.png)

datapythonista reviewed (approved):
lgtm, thanks!

datapythonista commented:
Unlucky this was not merged earlier.

You need to fix the conflict in a way that you get the lastest changes (from the conflicting PR in this case), and you make any fixes to that.

This PR is a diff of what you have and master, if you take your changes, you'll be deleting the work in that PR here.

I'm not sure of how the long set of units in that PR renders. If it doesn't render correctly, you can fix it here (instead of having the units in the type, just say `str` and then you show the units as a list in the description).

datapythonista commented:
Yes, if they were added that's for a reason. Avoid changing the content unless you really know what you're doing. But make the format look as good as possible. And in this case, I'm not sure how sphinx handles multilines in the types, we had problems with that in the past, so may be the current format is displayed broken.

ryankarlos commented:
@datapythonista Ive fixed the merge conflict to include changes from upstream. The units were not rendering properly when including in the type (the format was getting broken after the first line) so I shifted it to the description as you had suggested.


![screen shot 2018-11-13 at 00 10 16](https://user-images.githubusercontent.com/16509490/48382397-da9f5680-e6d8-11e8-89da-e94dd9e6654e.png)

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -16,31 +16,42 @@
-        - if False returns a np.timedelta64 or ndarray of values of dtype
-          timedelta64[ns]
+    arg : str, timedelta, list-like or Series
+          The argument which needs to be converted to timedelta.
```
`The argument` sounds a bit confusing to me. I'd say something like `The data to be converted to timedelta.` or something similar.

The indentation of the description doesn't seem right, should be 4 spaces respect the parameter name. Same in the next sections.

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -16,31 +16,42 @@
+    arg : str, timedelta, list-like or Series
+          The argument which needs to be converted to timedelta.
+    unit : str, default 'ns'
+         ('Y', 'M', 'W', 'D', 'days', 'day', 'hours', hour', 'hr',
```
It's good to have the possible values, but we do want a description for this parameter too. Explaining what it is.

datapythonista commented on pandas/core/tools/timedeltas.py:
```diff
@@ -16,31 +16,42 @@
+         'nanosecond', 'N').
+    box : bool, default True
+          If True returns a Timedelta/TimedeltaIndex of the results.
+          If False returns a np.timedelta64 or ndarray of values of dtype
```
I think this will render better as a list, as it was before. Also, I wouldn't abbreviate numpy in the description, and we should probably use it for `numpy.ndarray`.

datapythonista reviewed (changes requested):
Sorry for the delay in reviewing.

Looks good, just couple of minor things, and I think it's ready to be merged.

ryankarlos committed:
Additional small fixes

jreback commented:
@datapythonista over to you. merge when satisfied.
